### PR TITLE
AF-1817: Gzip filter

### PR DESCRIPTION
**Pull request by tiagobento**
Adds a filter that can compress a servlet response as Gzip. Please see a detailed explanation on the related kiegroup/kie-wb-distributions PR.

Related PRs:
https://github.com/kiegroup/appformer/pull/631
https://github.com/kiegroup/kie-wb-distributions/pull/877
https://github.com/kiegroup/kie-docs/pull/1395

### CODE REVIEW

File: uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/util/gzip/GzipFilter.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+    public void init(final FilterConfig filterConfig) {
+        // Empty on purpose
+    }
+
+    public void doFilter(final ServletRequest req,
+                         final ServletResponse res,
+                         final FilterChain chain) throws IOException, ServletException {
+
+        if (!(req instanceof HttpServletRequest)) {
```
Comment (porcelli) [inline uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/util/gzip/GzipFilter.java]: please add an environment variable to enable/disable this feature. the expected default is enabled.

and please add this env variable to kie-docs too.

Also: tests are missing :)

**Comment by porcelli** [comment]
still missing unit tests

**Comment by tiagobento** [comment]
@porcelli System property, tests and kie-docs PR were added.

**Comment by domhanak** [comment]
jenkins retest this

**Comment by tiagobento** [comment]
Jenkins, please execute full downstream build

**Comment by domhanak** [comment]
Roses are red,
violets are blue,
Jenkins execute full downstream build!

**Comment by domhanak** [comment]
When I run QE smoke tests , I can see this exception; Trying to find the root cause now - so far it seems that it happens when you create a legacy designer process  or migrate current stunner process to legacy and validation kick in. I will let you know when I know the root cause.

```12:58:05,970 ERROR [stderr] (default task-21) java.io.IOException: Output stream already closed.
12:58:05,971 ERROR [stderr] (default task-21) 	at org.uberfire.backend.server.util.gzip.GzipResponseServletOutputStream.flush(GzipResponseServletOutputStream.java:68)
12:58:05,971 ERROR [stderr] (default task-21) 	at io.undertow.servlet.spec.RequestDispatcherImpl.forwardImpl(RequestDispatcherImpl.java:237)
12:58:05,971 ERROR [stderr] (default task-21) 	at io.undertow.servlet.spec.RequestDispatcherImpl.forwardImplSetup(RequestDispatcherImpl.java:149)
12:58:05,977 ERROR [stderr] (default task-21) 	at io.undertow.servlet.spec.RequestDispatcherImpl.forward(RequestDispatcherImpl.java:111)
12:58:05,977 ERROR [stderr] (default task-21) 	at org.jbpm.designer.filter.DesignerResourcesRedirectFilter.doFilter(DesignerResourcesRedirectFilter.java:50)
12:58:05,977 ERROR [stderr] (default task-21) 	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
12:58:05,977 ERROR [stderr] (default task-21) 	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
12:58:05,977 ERROR [stderr] (default task-21) 	at org.uberfire.backend.server.util.gzip.GzipFilter.doFilter(GzipFilter.java:58)
12:58:05,977 ERROR [stderr] (default task-21) 	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
12:58:05,977 ERROR [stderr] (default task-21) 	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
12:58:05,977 ERROR [stderr] (default task-21) 	at org.uberfire.ext.security.server.SecureHeadersFilter.doFilter(SecureHeadersFilter.java:110)
12:58:05,977 ERROR [stderr] (default task-21) 	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
12:58:05,977 ERROR [stderr] (default task-21) 	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
12:58:05,977 ERROR [stderr] (default task-21) 	at org.uberfire.ext.security.server.CacheHeadersFilter.doFilter(CacheHeadersFilter.java:73)
12:58:05,977 ERROR [stderr] (default task-21) 	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
12:58:05,977 ERROR [stderr] (default task-21) 	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
12:58:05,977 ERROR [stderr] (default task-21) 	at org.uberfire.ext.security.server.SecurityIntegrationFilter.doFilter(SecurityIntegrationFilter.java:70)
12:58:05,977 ERROR [stderr] (default task-21) 	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
12:58:05,978 ERROR [stderr] (default task-21) 	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
12:58:05,978 ERROR [stderr] (default task-21) 	at io.opentracing.contrib.jaxrs2.server.SpanFinishingFilter.doFilter(SpanFinishingFilter.java:55)
12:58:05,978 ERROR [stderr] (default task-21) 	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
12:58:05,978 ERROR [stderr] (default task-21) 	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
12:58:05,978 ERROR [stderr] (default task-21) 	at io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84)
12:58:05,978 ERROR [stderr] (default task-21) 	at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:62)
12:58:05,978 ERROR [stderr] (default task-21) 	at io.undertow.servlet.handlers.ServletChain$1.handleRequest(ServletChain.java:68)
12:58:05,978 ERROR [stderr] (default task-21) 	at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
12:58:05,978 ERROR [stderr] (default task-21) 	at org.wildfly.extension.undertow.security.SecurityContextAssociationHandler.handleRequest(SecurityContextAssociationHandler.java:78)
12:58:05,978 ERROR [stderr] (default task-21) 	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
12:58:05,978 ERROR [stderr] (default task-21) 	at io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:132)
12:58:05,978 ERROR [stderr] (default task-21) 	at io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)
12:58:05,978 ERROR [stderr] (default task-21) 	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
12:58:05,978 ERROR [stderr] (default task-21) 	at io.undertow.security.handlers.AuthenticationConstraintHandler.handleRequest(AuthenticationConstraintHandler.java:53)
12:58:05,978 ERROR [stderr] (default task-21) 	at io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)
12:58:05,978 ERROR [stderr] (default task-21) 	at io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:64)
12:58:05,979 ERROR [stderr] (default task-21) 	at io.undertow.servlet.handlers.security.ServletSecurityConstraintHandler.handleRequest(ServletSecurityConstraintHandler.java:59)
12:58:05,979 ERROR [stderr] (default task-21) 	at io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60)
12:58:05,979 ERROR [stderr] (default task-21) 	at io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77)
12:58:05,979 ERROR [stderr] (default task-21) 	at io.undertow.security.handlers.NotificationReceiverHandler.handleRequest(NotificationReceiverHandler.java:50)
12:58:05,979 ERROR [stderr] (default task-21) 	at io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43)
12:58:05,979 ERROR [stderr] (default task-21) 	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
12:58:05,979 ERROR [stderr] (default task-21) 	at org.wildfly.extension.undertow.security.jacc.JACCContextIdHandler.handleRequest(JACCContextIdHandler.java:61)
12:58:05,979 ERROR [stderr] (default task-21) 	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
12:58:05,979 ERROR [stderr] (default task-21) 	at org.wildfly.extension.undertow.deployment.GlobalRequestControllerHandler.handleRequest(GlobalRequestControllerHandler.java:68)
12:58:05,979 ERROR [stderr] (default task-21) 	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
12:58:05,979 ERROR [stderr] (default task-21) 	at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:292)
12:58:05,979 ERROR [stderr] (default task-21) 	at io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:81)
12:58:05,979 ERROR [stderr] (default task-21) 	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:138)
12:58:05,979 ERROR [stderr] (default task-21) 	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:135)
12:58:05,979 ERROR [stderr] (default task-21) 	at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
12:58:05,979 ERROR [stderr] (default task-21) 	at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
12:58:05,979 ERROR [stderr] (default task-21) 	at org.wildfly.extension.undertow.security.SecurityContextThreadSetupAction.lambda$create$0(SecurityContextThreadSetupAction.java:105)
12:58:05,979 ERROR [stderr] (default task-21) 	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1502)
12:58:05,979 ERROR [stderr] (default task-21) 	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1502)
12:58:05,979 ERROR [stderr] (default task-21) 	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1502)
12:58:05,979 ERROR [stderr] (default task-21) 	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1502)
12:58:05,980 ERROR [stderr] (default task-21) 	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:272)
12:58:05,980 ERROR [stderr] (default task-21) 	at io.undertow.servlet.handlers.ServletInitialHandler.access$000(ServletInitialHandler.java:81)
12:58:05,980 ERROR [stderr] (default task-21) 	at io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:104)
12:58:05,980 ERROR [stderr] (default task-21) 	at io.undertow.server.Connectors.executeRootHandler(Connectors.java:360)
12:58:05,980 ERROR [stderr] (default task-21) 	at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:830)
12:58:05,980 ERROR [stderr] (default task-21) 	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
12:58:05,980 ERROR [stderr] (default task-21) 	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1985)
12:58:05,980 ERROR [stderr] (default task-21) 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1487)
12:58:05,980 ERROR [stderr] (default task-21) 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1349)
12:58:05,980 ERROR [stderr] (default task-21) 	at java.lang.Thread.run(Thread.java:748)
12:58:05,980 ERROR [stderr] (default task-18) java.io.IOException: Output stream already closed.

```

**Comment by domhanak** [comment]
@tiagobento @ederign @mbiarnes Any1 has an idea what is wrong with compile downstream build?

File: uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/util/gzip/GzipHttpServletResponseWrapper.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            throw new IllegalStateException("getWriter() was already called");
+        }
+
+        stream = createOutputStream();
+        writer = new PrintWriter(new OutputStreamWriter(stream, StandardCharsets.UTF_8));
+        return writer;
+    }
+
+    @Override
+    public void setContentLength(final int length) {
```
Comment (domhanak) [inline uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/util/gzip/GzipHttpServletResponseWrapper.java]: empty on purpose?

**Comment by tiagobento** [comment]
@domhanak Do you have a reproducer for that stack you posted?

File: uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/util/gzip/GzipHttpServletResponseWrapper.java (hunk 2 above)
Comment (tiagobento) [reply uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/util/gzip/GzipHttpServletResponseWrapper.java]: Yes

**Comment by tiagobento** [comment]
@domhanak Looks like compile downstream is fixed :)

**Comment by domhanak** [comment]
Let's do the FDB once more.

Roses are red,
violets are blue,
Jenkins execute full downstream build!

**Comment by porcelli** [comment]
the fdb build failed. However the list of failed test are all covered by http://janhrcek.cz/random-failures/ report

so makes sense to merge this

**Comment by tiagobento** [comment]
@porcelli I was just writing the exact same thing. Thanks :)

**Pull request merged**